File: 3-cloudrun-golang/controller/device_controller.go
package controller

import (
	"encoding/json"
	"net/http"
	"strings"

	"github.com/benchmark/cloudrun-go/model"
	"github.com/benchmark/cloudrun-go/service"
)

type DeviceController struct {
	service *service.DeviceService
}

func NewDeviceController(svc *service.DeviceService) *DeviceController {
	return &DeviceController{service: svc}
}

func (c *DeviceController) RegisterRoutes(mux *http.ServeMux) {
	mux.HandleFunc("POST /api/devices", c.Create)
	mux.HandleFunc("GET /api/devices/{id}", c.GetByID)
	mux.HandleFunc("GET /api/devices", c.List)
}

func (c *DeviceController) Create(w http.ResponseWriter, r *http.Request) {
	var req model.CreateDeviceRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, `{"error":"invalid request body"}`, http.StatusBadRequest)
		return
	}

	device := c.service.Create(req)

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(device)
}

func (c *DeviceController) GetByID(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	if id == "" {
		http.Error(w, `{"error":"id required"}`, http.StatusBadRequest)
		return
	}

	device, found := c.service.GetByID(id)
	if !found {
		http.Error(w, `{"error":"not found"}`, http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(device)
}

func (c *DeviceController) List(w http.ResponseWriter, r *http.Request) {
	status := strings.TrimSpace(r.URL.Query().Get("status"))
	devices := c.service.ListByStatus(status)

	if devices == nil {
		devices = []model.Device{}
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(devices)
}

File: 1b-gke-spring-jvm/src/main/java/com/benchmark/device/controller/DeviceController.java
package com.benchmark.device.controller;

import com.benchmark.device.model.Device;
import com.benchmark.device.service.DeviceService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/api/devices")
public class DeviceController {

    private final DeviceService service;

    public DeviceController(DeviceService service) {
        this.service = service;
    }

    @PostMapping
    public ResponseEntity<Device> create(@RequestBody Device device) {
        Device created = service.create(device);
        return ResponseEntity.status(201).body(created);
    }

    @GetMapping("/{id}")
    public ResponseEntity<Device> getById(@PathVariable UUID id) {
        return service.getById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping
    public List<Device> list(@RequestParam(required = false) String status) {
        return service.listByStatus(status);
    }
}
